pkg/queue/concurrency_state.go

```diff
@@ -0,0 +1,113 @@
+/*
+Copyright 2021 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package queue
+
+import (
+	"net/http"
+
+	"go.uber.org/zap"
+)
+
+// ConcurrencyStateHandler tracks the in flight requests for the pod. When the requests
+// drop to zero, it runs the `pause` function, and when requests scale up from zero, it
+// runs the `resume` function. If either of `pause` or `resume` are not passed, it runs
+// the respective local function(s). The local functions are the expected behavior; the
+// function parameters are enabled primarily for testing purposes.
+func ConcurrencyStateHandler(endpoint string, h http.Handler, logger *zap.SugaredLogger, pause, resume func()) http.HandlerFunc {
+	logger.Info("Concurrency state tracking enabled")
+
+	if pause == nil {
+		pause = func() {
+			logger.Info("Requests dropped to zero, pausing pod ...")
+			pauseEndpoint := endpoint + ":9696/freeze"
```

Comment:
should we hard-code the port and path here or let the endpoint env var define that?

Reply:
Port I think is fine to include in the envvar. If we include path in the envvar, then I think we'd need to pass freeze/resume as part of the request (or use two endpoints, but I don't think that's a good idea). 

Not sure which option makes more sense (keeping path hard coded or adding the action to the request) makes more sense. In either case, I think we'll ultimately have to define what is required from any custom "pod freezer" service that gets called out to.

Reply:
if we add the path then I think we probably want to be more generic that freeze and thaw. Another option though would be to use a single endpoint and pass a json structure to it (so it would be the same endpoint either way, but the json structure would define whether the event was to-zero or from-zero).

Reply:
Thanks. The more I think about it, a single endpoint with the action taken in the request itself seems like the right path to take.

Reply:
@julz Another thought -- if the path is in the envvar, then we don't necessarily need separate pause/resume functions. We could use a single function and pass the JSON as a parameter. 

If that approach makes sense, then perhaps I should update the function signature in #11783 before it merges so that it stays consistent...

Reply:
I think we may still want it as separate functions to make it easy to test (otherwise you'd have to assert on the json being sent), but we can always refactor as we play with this more. We could also have a single function with a boolean or something. Probably adding some unit tests would be helpful at this point anyway, and then I can stop guessing about what will or won't make things more testable ;).

Reply:
It was writing the tests that triggered my thought (and also led me to the same conclusion you have, that it's easier to test with separate functions) :)